cpp/src/arrow/compute/kernels/minmax.cc

```diff
@@ -40,19 +40,35 @@ struct MinMaxState<ArrowType, enable_if_integer<ArrowType>> {
 
   ThisType& operator+=(const ThisType& rhs) {
     this->has_nulls |= rhs.has_nulls;
-    this->min = std::min(this->min, rhs.min);
-    this->max = std::max(this->max, rhs.max);
+    if (ARROW_PREDICT_TRUE(this->initialized && rhs.initialized)) {
+      this->min = std::min(this->min, rhs.min);
+      this->max = std::max(this->max, rhs.max);
+    } else if (!this->initialized && rhs.initialized) {
+      this->min = rhs.min;
+      this->max = rhs.max;
+      this->initialized = true;
+    }
     return *this;
   }
 
   void MergeOne(c_type value) {
-    this->min = std::min(this->min, value);
-    this->max = std::max(this->max, value);
```

Comment:
This simple code is actually very efficient. std::min/max uses cpu conditional move instructions which are branchless. And compiler can vectorize the hot loop with simd instructions. See https://godbolt.org/z/ATEdEs.
I suspect this textbook trick may not work as expected in practice.
Anyway, concrete benchmark data will be the best reference.

Reply:
I agree with this. Instead of rewriting code, how about adding `inline` to encourage vectorization in a loop?

Reply:
The linked godbolt is using vectorized instruction, so does enabling [avx2](https://godbolt.org/z/G7Ppjw). I also checked locally via objdump:

```asm
  afa708:       0f 84 31 01 00 00       je     afa83f <arrow::compute::MinMaxAggregateFunction<arrow::UInt32Type>::Consume(arrow::Array const&, arrow::compute::MinMaxState<arrow::UInt32Type, void>*) const+0x2af>
  afa70e:       48 29 d7                sub    %rdx,%rdi
  afa711:       4b 8d 1c ac             lea    (%r12,%r13,4),%rbx
  afa715:       48 83 c3 60             add    $0x60,%rbx
  afa719:       c5 fd 76 c0             vpcmpeqd %ymm0,%ymm0,%ymm0
  afa71d:       c5 f1 ef c9             vpxor  %xmm1,%xmm1,%xmm1
  afa721:       31 f6                   xor    %esi,%esi
  afa723:       c5 e9 ef d2             vpxor  %xmm2,%xmm2,%xmm2
  afa727:       c5 e5 76 db             vpcmpeqd %ymm3,%ymm3,%ymm3
  afa72b:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
  afa730:       c5 fe 6f 64 b3 a0       vmovdqu -0x60(%rbx,%rsi,4),%ymm4
  afa736:       c5 fe 6f 6c b3 c0       vmovdqu -0x40(%rbx,%rsi,4),%ymm5
  afa73c:       c5 fe 6f 74 b3 e0       vmovdqu -0x20(%rbx,%rsi,4),%ymm6
  afa742:       c5 fe 6f 3c b3          vmovdqu (%rbx,%rsi,4),%ymm7
  afa747:       c4 e2 7d 3b c4          vpminud %ymm4,%ymm0,%ymm0
  afa74c:       c4 e2 65 3b dd          vpminud %ymm5,%ymm3,%ymm3
  afa751:       c4 e2 75 3f cc          vpmaxud %ymm4,%ymm1,%ymm1
  afa756:       c4 e2 6d 3f d5          vpmaxud %ymm5,%ymm2,%ymm2
  afa75b:       c4 e2 7d 3b c6          vpminud %ymm6,%ymm0,%ymm0
  afa760:       c4 e2 65 3b df          vpminud %ymm7,%ymm3,%ymm3
  afa765:       c4 e2 75 3f ce          vpmaxud %ymm6,%ymm1,%ymm1
  afa76a:       c4 e2 6d 3f d7          vpmaxud %ymm7,%ymm2,%ymm2
  afa76f:       48 83 c6 20             add    $0x20,%rsi
  afa773:       48 83 c7 fe             add    $0xfffffffffffffffe,%rdi
  afa777:       75 b7                   jne    afa730 <arrow::compute::MinMaxAggregateFunction<arrow::UInt32Type>::Consume(arrow::Array const&, arrow::compute::MinMaxState<arrow::UInt32Type, void>*) const+0x1a0>
  afa779:       48 85 d2                test   %rdx,%rdx
  afa77c:       4c 8b 2c 24             mov    (%rsp),%r13

Reply:
Side note, gcc cannot auto-vectorize this until gcc-10.